issue-2253

schlenkibus opened the pull request:
added check if preset message changed before sending same message again. closes #2253

- [x] todo test

MatthiasSeeber reviewed (changes requested):
> Preset Messages müssen versendet werden, wenn: der User ein Preset oder einen Part lädt (Recall) oder einen Sound konvertiert/randomisiert/initialisiert.
> Sie müssen/sollten nicht versendet werden, wenn ein Sound als Preset abgespeichert wird (Store append/overwrite/insert).

schlenkibus commented:
I extracted the use cases for Store Actions from all over the codebase into one central location, now store actions called from the webui and the boled will run the same code. The original issue of sending preset messages when stored is now fixed. Every store will update the selection and loaded preset status to the just inserted, appended preset.

MatthiasSeeber commented on projects/epc/playground/src/presets/Bank.cpp:
```diff
@@ -435,7 +436,7 @@ Preset *Bank::prependPreset(UNDO::Transaction *transaction, std::unique_ptr<Pres
 {
   auto newPreset = prependPreset(transaction, std::move(preset));
-  getEditBuffer()->undoableLoad(transaction, newPreset);
+  getEditBuffer()->undoableLoad(transaction, newPreset, false);
```
I noticed that in [426, 453], the `getPresetManager()->onPresetStored(newPreset);` is invoked. Why isn't it called here?

hhoegelo replied:
Good question, I'd like to know that, too. @schlenkibus ?

MatthiasSeeber commented on projects/epc/playground/src/proxies/audio-engine/AudioEngineProxy.h:
```diff
@@ -31,24 +31,27 @@ class AudioEngineProxy
+
   void freezeParameterMessages();
-  void thawParameterMessages();
+  void thawParameterMessages(bool send);
```
The name is rather confusing to me. What does `thaw` mean?

hhoegelo replied:
It is just the opposite of 'freeze' - maybe 'pause' and 'resume' would be more appropriate words. 
Pausing the hundreds of parameter messages makes sense, if we know that eventually there will be a full preset message anyway. 
The wording is borrowed from glib: https://developer.gnome.org/glibmm/stable/classGlib_1_1ObjectBase.html#a1bd8ea7bd8c4084ade6b3c27dddf06a4

MatthiasSeeber reviewed (approved):
I observed that Presets are only sent to the AudioEngine only when needed - all Store process are "mute" now. Out of curiosity, I have two additional questions.

hhoegelo commented on projects/epc/playground/src/presets/PresetManager.cpp:
```diff
@@ -999,3 +999,13 @@ void PresetManager::autoLoadPresetAccordingToLoadType()
+
+void PresetManager::onPresetStored(const Preset *storedPreset)
+{
+  m_presetStoreHappened.deferedSend(storedPreset);
```
It is not actually a problem, because due to our undo logic a preset will 'never' be deleted, but I don't like the idea of deferred signaling of an object by pointer. Theoretically it is possible that the preset is already deleted at the time the signal notification eventually happens. Also, the - so far - one and only user of the signal doesn't care about the preset pointer. So I suggest to remove the pointer and just signal that "some preset store has happened".

schlenkibus replied:
makes sense, i will implement your suggestion
